### PR TITLE
Emscripten Build

### PR DESCRIPTION
**Pull request by CDanU**
see issue #796

### CODE REVIEW

File: src/emscripten/libUncrustify.d.ts [hunk 1]
```diff
@@ -0,0 +1,823 @@
+interface EmscriptenVector< T >
```
Comment (mihaipopescu) [inline src/emscripten/libUncrustify.d.ts]: Is this generated ?

File: src/options.h [hunk 2]
```diff
@@ -843,15 +843,15 @@ struct option_map_value
    argtype_e          type;
    int                min_val;
    int                max_val;
-   const char         *name;
-   const char         *short_desc;
-   const char         *long_desc;
+   string             name;
```
Comment (mihaipopescu) [inline src/options.h]: Why do you really need this as a string while the data is static ?

**Comment by mihaipopescu** [comment]
We may have to hold on as this is quite a consistent change and we need time to digest all this.

File: src/uncrustify_emscripten.cpp [hunk 3]
```diff
@@ -0,0 +1,1312 @@
...
+* @return formated string
+*/
+string uncrustify( string file )
+{
+    // overload for default frag parameter
+    return uncrustify( file, false );
+}
+
+
+EMSCRIPTEN_BINDINGS( MainModule )
```
Comment (CDanU) [inline src/uncrustify_emscripten.cpp]: see above

File: src/emscripten/libUncrustify.d.ts (hunk 1 above)
Comment (CDanU) [inline src/emscripten/libUncrustify.d.ts]: This is a declaration file for Typescript. It helps working with the generated Emscripten library in a Typescrript environment. 

It is was also created by hand and the enum values should also be generated at some point.

File: src/options.h (hunk 2 above)
Comment (CDanU) [inline src/options.h]: It is not really needed to work with strings here but it makes everything so much easier when one can avoiding pointers in emscripten.

**Comment by gmaurel** [comment]
Why do you need the changes?

**Comment by CDanU** [comment]
Could you specify which changes? 

- 827362c, 1dae768 and 6542fce are just some random improvements I had made, those are not needed but improve readability
- 93c9a73 Since I cannot work with the C FILE type, this allows me to parse the config file line by line as a string
- cdae69b I don't see why option_name_map is using a string as index, the enum value should be more suitable here and it also allows an better/easier interface for the js lib
- 3f358d1  Here also I don't get why a list is used, a vector should be more suitable and Emscripten has already bindings for the vector structure implemented.
- 3f0022e like 93c9a73 this is used to read a passed config file as a string, although I guess I could move that function into the Emscripten interface file
- ce2a89a will be replaced with something non intrusive

Everything else is either part of the Emscripten interface file or removes static specifiers in order to allow the Emscripten interface file to use functions.

**Comment by gmaurel** [comment]
Sorry, wrong question.
I try again:
What is the relation uncrustify <--> emscripten

**Pull request closed without merging**